### ISSUE TITLE
Tutorial 2 - Build a proof of existence dApp

### ISSUE THREAD

**Issue by Inux131** [paritytech/substrate#10999]
Hi,

may you help?

- I Added a skeleton set of pallet dependencies and macros as described and replaced the macros with the given code lines. 
- cargo check -p node-template-runtime -> Gives me the following errors:

`robert@robert:~/substrate-node-template/pallets/template/src$ cargo check -p node-template-runtime
   Compiling node-template-runtime v4.0.0-dev (/home/robert/substrate-node-template/runtime)
    Checking pallet-template v4.0.0-dev (/home/robert/substrate-node-template/pallets/template)
error[E0599]: the function or associated item `contains_key` exists for struct `frame_support::pallet_prelude::StorageMap<_GeneratedPrefixForStorageProofs<T>, frame_support::Blake2_128Concat, Vec<u8>, (<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber), frame_support::pallet_prelude::ValueQuery>`, but its trait bounds were not satisfied
   --> pallets/template/src/lib.rs:75:31
    |
75  |         ensure!(!Proofs::<T>::contains_key(&proof), Error::<T>::ProofAlreadyClaimed);
    |                               ^^^^^^^^^^^^ function or associated item cannot be called on `frame_support::pallet_prelude::StorageMap<_GeneratedPrefixForStorageProofs<T>, frame_support::Blake2_128Concat, Vec<u8>, (<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber), frame_support::pallet_prelude::ValueQuery>` due to unsatisfied trait bounds
    |
   ::: /home/robert/.cargo/git/checkouts/substrate-7e08433d4c370a21/22d40c7/frame/support/src/storage/types/mod.rs:89:1
    |
89  | pub struct ValueQuery;
    | ---------------------- doesn't satisfy `_: QueryKindTrait<(<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber), GetDefault>`
    |
   ::: /home/robert/.cargo/git/checkouts/substrate-7e08433d4c370a21/22d40c7/frame/support/src/traits/misc.rs:420:1
    |
420 | pub struct GetDefault;
    | ---------------------- doesn't satisfy `_: frame_support::traits::Get<(<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber)>`
    |
    = note: the following trait bounds were not satisfied:
            `frame_support::pallet_prelude::ValueQuery: QueryKindTrait<(<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber), GetDefault>`
            `GetDefault: frame_support::traits::Get<(<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber)>`

error[E0599]: the function or associated item `insert` exists for struct `frame_support::pallet_prelude::StorageMap<_GeneratedPrefixForStorageProofs<T>, frame_support::Blake2_128Concat, Vec<u8>, (<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber), frame_support::pallet_prelude::ValueQuery>`, but its trait bounds were not satisfied
   --> pallets/template/src/lib.rs:81:22
    |
81  |         Proofs::<T>::insert(&proof, (&sender, current_block));
    |                      ^^^^^^ function or associated item cannot be called on `frame_support::pallet_prelude::StorageMap<_GeneratedPrefixForStorageProofs<T>, frame_support::Blake2_128Concat, Vec<u8>, (<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber), frame_support::pallet_prelude::ValueQuery>` due to unsatisfied trait bounds
`

**Comment by ggwpez** [comment]
The `AccountId` that is stored in the map is not defaultable anymore, could this but out the tutorial here?

**Comment by Inux131** [comment]
Hi,
how can I set the AccountID to nondefaultable?

**Comment by ggwpez** [comment]
If you remove the `ValueQuery` from the end of the `StorageMap` definition, but then a lot of other things are probably going to break.  

@bkchr should now if this is indeed a problem with the tutorial.

**Comment by bkchr** [comment]
Yes, you need to use `OptionQuery`. Then it will not require anymore that the type implements `Default`

**Comment by Inux131** [comment]
I changed ValueQuery to OptionQuery.

now the following error is occurring. May you please help again?

`cargo check -p node-template-runtime
   Compiling node-template-runtime v4.0.0-dev (/home/robert/substrate-node-template/runtime)
    Checking pallet-template v4.0.0-dev (/home/robert/substrate-node-template/pallets/template)
error[E0308]: mismatched types
   --> pallets/template/src/lib.rs:102:13
    |
102 |         let (owner, _) = Proofs::<T>::get(&proof);
    |             ^^^^^^^^^^   ------------------------ this expression has type `std::option::Option<(<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber)>`
    |             |
    |             expected enum `std::option::Option`, found tuple
    |
    = note: expected enum `std::option::Option<(<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber)>`
              found tuple `(_, _)`

error[E0277]: the trait bound `Vec<u8>: MaxEncodedLen` is not satisfied
  --> pallets/template/src/lib.rs:46:15
   |
46 |     #[pallet::pallet]
   |               ^^^^^^ the trait `MaxEncodedLen` is not implemented for `Vec<u8>`
   |
   = note: required because of the requirements on the impl of `StorageInfoTrait` for `frame_support::pallet_prelude::StorageMap<_GeneratedPrefixForStorageProofs<T>, frame_support::Blake2_128Concat, Vec<u8>, (<T as frame_system::Config>::AccountId, <T as frame_system::Config>::BlockNumber)>`

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `pallet-template` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
`
